include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -641,6 +675,120 @@ namespace internal
+          shape_info.data.front().shape_gradients.data();
+        for (unsigned int c = 0; c < n_components; ++c)
+          for (unsigned int d = 0; d < dim; ++d)
+            for (unsigned int q = 0; q < n_q_points; ++q)
+              {
+                gradients_quad[n_q_points * dim * c + n_q_points * d + q] = 0.0;
+                for (unsigned int i = 0; i < n_dofs; ++i)
+                  gradients_quad[n_q_points * dim * c + n_q_points * d + q] +=
+                    shape_gradients[d * n_dofs * n_q_points + i * n_q_points +
+                                    q] *
+                    values_dofs_actual[n_dofs * dim * c + i];
+              }
+      }
+
+    if (evaluation_flag & EvaluationFlags::hessians)
+      {
+        Assert(false, ExcNotImplemented());
+        (void)hessians_quad;
+      }
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+  inline void
+  FEEvaluationImpl<
+    MatrixFreeFunctions::tensor_none,
+    dim,
+    fe_degree,
+    n_q_points_1d,
+    Number>::integrate(const unsigned int                     n_components,
+                       const EvaluationFlags::EvaluationFlags integration_flag,
+                       const MatrixFreeFunctions::ShapeInfo<Number> &shape_info,
+                       Number *   values_dofs_actual,
+                       Number *   values_quad,
+                       Number *   gradients_quad,
+                       Number *   scratch_data,
+                       const bool add_into_values_array)
+  {
+    (void)scratch_data;
+
+    const unsigned int n_dofs     = shape_info.dofs_per_component_on_cell;
+    const unsigned int n_q_points = shape_info.n_q_points;
+
+    if (add_into_values_array == false)
+      for (unsigned int i = 0; i < n_components * n_dofs; ++i)
+        values_dofs_actual[i] = 0.0;
+
+    if (integration_flag & EvaluationFlags::values)
+      {
+        const auto shape_values = shape_info.data.front().shape_values.data();
+        for (unsigned int c = 0; c < n_components; ++c)
+          for (unsigned int q = 0; q < n_q_points; ++q)
+            for (unsigned int i = 0; i < n_dofs; ++i)
+              values_dofs_actual[n_dofs * c + i] +=
+                shape_values[i * n_q_points + q] *
+                values_quad[n_q_points * c + q];
+      }
+
+    if (integration_flag & EvaluationFlags::gradients)
+      {
+        const auto shape_gradients =
+          shape_info.data.front().shape_gradients.data();
+        for (unsigned int c = 0; c < n_components; ++c)
+          for (unsigned int d = 0; d < dim; ++d)
+            for (unsigned int q = 0; q < n_q_points; ++q)
+              for (unsigned int i = 0; i < n_dofs; ++i)
+                values_dofs_actual[n_dofs * dim * c + i] +=
+                  shape_gradients[d * n_dofs * n_q_points + i * n_q_points +
+                                  q] *
+                  gradients_quad[n_q_points * dim * c + n_q_points * d + q];
+      }
+  }
```

Comment:
@kronbichler What I find "magical" about the current version that everything "simplex" related is centralized and once we should decide to work on more efficient interpolation routines it is clear what to touch!

Reply:
That's indeed very nice!